packages/server/lib/server-base.ts

```diff
@@ -174,6 +164,37 @@ export abstract class ServerBase<TSocket extends SocketE2E | SocketCt> {
     return this.ensureProp(this._httpsProxy, 'open')
   }
 
+  setupMultiDomain () {
+    this._eventBus.on('cross:domain:delaying:html', (request) => {
+      this.socket.localBus.once('ready:for:domain', (args) => {
+        this._eventBus.emit('ready:for:domain', args)
+      })
+
+      this.socket.toDriver('cross:domain:delaying:html', request)
+    })
+
+    // This is handled separately from above since we may not need to delay the html
+    // when we are in the scenario where the user creates the switchToDomain first
+    // and then visits or navigates within the secondary domain
+    this.socket.localBus.on('ready:for:domain', ({ originPolicy, failed }) => {
+      if (failed) return
+
+      const existingOrigin = this.remoteStates.get(originPolicy)
+
+      // since this is just the switchToDomain starting, we don't want to override
+      // the existing origin if it already exists
+      if (!existingOrigin) {
+        this.remoteStates.set(originPolicy, { isMultiDomain: true })
+      }
+
+      this.remoteStates.addOrigin(originPolicy)
+    })
+
+    this.socket.localBus.on('cross:origin:finished', () => {
+      this.remoteStates.removeCurrentOrigin()
+    })
+  }
```

Comment:
I believe you mentioned looking to set up these event listeners within the `RemoteStates` class, and I agree that would be ideal. Maybe just have a method on `RemoteStates` that you call right here with `this.socket`?

Reply:
Added RemoteStates class. [556a615](https://github.com/cypress-io/cypress/pull/20752/commits/556a615bcc29fac5f31435c292f72575a73e8b19)